File: src/test/java/com/example/the_machine/operator/registry/SemanticSearchOperatorRegistryTest.java
package com.example.the_machine.operator.registry;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.lenient;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.example.the_machine.domain.OperatorEmbedding;
import com.example.the_machine.operator.Operator;
import com.example.the_machine.operator.ToolSpec;
import com.example.the_machine.repo.OperatorEmbeddingRepository;
import com.example.the_machine.service.EmbeddingService;
import com.example.the_machine.service.HashCalculationService;
import java.util.List;
import java.util.Optional;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

/**
 * Unit tests for SemanticSearchOperatorRegistry functionality including operator discovery,
 * embedding generation, and semantic search capabilities.
 */
@ExtendWith(MockitoExtension.class)
class SemanticSearchOperatorRegistryTest {

  @Mock
  private OperatorEmbeddingRepository embeddingRepository;

  @Mock
  private EmbeddingService embeddingService;

  @Mock
  private HashCalculationService hashCalculationService;

  @Mock
  private Operator mockOperator;

  @Mock
  private ToolSpec mockToolSpec;

  private SemanticSearchOperatorRegistry registry;

  private final float[] testEmbedding = {0.1f, 0.2f, 0.3f, 0.4f, 0.5f};
  private final String operatorName = "test-operator";
  private final String operatorDescription = "Test operator for unit testing";

  @BeforeEach
  void setUp() {
    registry = new SemanticSearchOperatorRegistry(embeddingRepository, embeddingService,
        hashCalculationService);

    // Setup common mock behavior with lenient to avoid unnecessary stubbing exceptions
    lenient().when(embeddingService.generateEmbedding(any(String.class)))
        .thenReturn(testEmbedding);
    lenient().when(hashCalculationService.calculateSha256Hash(any(String.class)))
        .thenReturn("mock-hash-12345");

    // Setup operator and tool spec mocks
    lenient().when(mockOperator.spec()).thenReturn(mockToolSpec);
    lenient().when(mockToolSpec.getName()).thenReturn(operatorName);
    lenient().when(mockToolSpec.getDescription()).thenReturn(operatorDescription);
    lenient().when(mockToolSpec.getInputKeys()).thenReturn(java.util.Set.of("input1", "input2"));
    lenient().when(mockToolSpec.getOutputKeys()).thenReturn(java.util.Set.of("output1"));
  }

  @Test
  void testGetOperatorByName() {
    // Given
    registry.getOperatorsByName().put(operatorName, mockOperator);

    // When
    val result = registry.get(operatorName);

    // Then
    assertThat(result).isPresent();
    assertThat(result.get()).isEqualTo(mockOperator);
  }

  @Test
  void testGetOperatorByNameNotFound() {
    // When
    val result = registry.get("non-existent-operator");

    // Then
    assertThat(result).isEmpty();
  }

  @Test
  void testSearchByTaskDescription() {
    // Given
    val taskDescription = "process some text";
    val operatorEmbedding = OperatorEmbedding.builder()
        .operatorName(operatorName)
        .embedding(testEmbedding)
        .build();

    registry.getOperatorsByName().put(operatorName, mockOperator);

    when(embeddingRepository.findTopByEmbeddingSimilarity(any(float[].class), any(Pageable.class)))
        .thenReturn(List.of(operatorEmbedding));

    // When
    val results = registry.searchByTask(taskDescription);

    // Then
    assertThat(results).hasSize(1);
    assertThat(results.getFirst()).isEqualTo(mockToolSpec);
    verify(embeddingService).generateEmbedding(taskDescription);
    verify(embeddingRepository).findTopByEmbeddingSimilarity(eq(testEmbedding),
        eq(PageRequest.of(0, 3)));
  }

  @Test
  void testSearchByTaskDescriptionWithMaxResults() {
    // Given
    val taskDescription = "process some text";
    val operatorEmbedding = OperatorEmbedding.builder()
        .operatorName(operatorName)
        .embedding(testEmbedding)
        .build();

    registry.getOperatorsByName().put(operatorName, mockOperator);

    when(embeddingRepository.findTopByEmbeddingSimilarity(any(float[].class), any(Pageable.class)))
        .thenReturn(List.of(operatorEmbedding));

    // When
    val results = registry.searchByTask(taskDescription);

    // Then
    assertThat(results).hasSize(1);
    assertThat(results.getFirst()).isEqualTo(mockToolSpec);
    verify(embeddingRepository).findTopByEmbeddingSimilarity(eq(testEmbedding),
        eq(PageRequest.of(0, 3)));
  }

  @Test
  void testSearchByTaskHandlesException() {
    // Given
    val taskDescription = "process some text";
    when(embeddingService.generateEmbedding(taskDescription))
        .thenThrow(new RuntimeException("Embedding service unavailable"));

    // When
    val results = registry.searchByTask(taskDescription);

    // Then
    assertThat(results).isEmpty();
  }

  @Test
  void testCreateEmbeddingText() {
    // Given - using package-private method for testing
    val embeddingText = registry.createEmbeddingTextForTesting(mockToolSpec);

    // Then - check individual components since Set order is not guaranteed
    assertThat(embeddingText)
        .contains("Tool: " + operatorName)
        .contains(operatorDescription)
        .contains("Inputs:")
        .contains("input1")
        .contains("input2")
        .contains("Outputs: output1");
  }

  @Test
  void testInitializeWithExistingEmbedding() {
    // Given
    lenient().when(embeddingRepository.findByOperatorName(operatorName))
        .thenReturn(Optional.of(OperatorEmbedding.builder().build()));

    // Note: Since we can't easily mock ServiceLoader, this test would need integration testing
    // or a more complex setup. For now, we verify the behavior when embedding already exists.

    // This test mainly verifies that the method doesn't throw exceptions
    // The actual ServiceLoader testing would be done in integration tests

    // When - no action needed as ServiceLoader is difficult to mock in unit tests
    // Then - no assertion needed, just ensuring no exceptions are thrown
  }
}

File: src/test/kotlin/com/example/the_machine/operator/registry/SemanticSearchOperatorRegistryTest.kt
package com.example.the_machine.operator.registry

import com.example.the_machine.domain.OperatorEmbedding
import com.example.the_machine.operator.Operator
import com.example.the_machine.operator.ToolSpec
import com.example.the_machine.repo.OperatorEmbeddingRepository
import com.example.the_machine.service.EmbeddingService
import com.example.the_machine.service.HashCalculationService
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import java.util.*

/**
 * Unit tests for SemanticSearchOperatorRegistry functionality including operator discovery,
 * embedding generation, and semantic search capabilities.
 */
@ExtendWith(MockKExtension::class)
class SemanticSearchOperatorRegistryTest {

  private val embeddingRepository: OperatorEmbeddingRepository = mockk()
  private val embeddingService: EmbeddingService = mockk()
  private val hashCalculationService: HashCalculationService = mockk()
  private val mockOperator: Operator = mockk(relaxed = true)
  private val mockToolSpec: ToolSpec = mockk(relaxed = true)

  private lateinit var registry: SemanticSearchOperatorRegistry

  private val testEmbedding = floatArrayOf(0.1f, 0.2f, 0.3f, 0.4f, 0.5f)
  private val operatorName = "test-operator"
  private val operatorDescription = "Test operator for unit testing"

  @BeforeEach
  fun setUp() {
    registry = SemanticSearchOperatorRegistry(
      embeddingRepository,
      embeddingService,
      hashCalculationService
    )

    // Setup common mock behavior using MockK's every blocks
    every { embeddingService.generateEmbedding(any<String>()) } returns testEmbedding
    every { hashCalculationService.calculateSha256Hash(any<String>()) } returns "mock-hash-12345"

    // Setup operator and tool spec mocks - using relaxed mocks so these are optional
    every { mockOperator.spec() } returns mockToolSpec
    every { mockToolSpec.name } returns operatorName
    every { mockToolSpec.description } returns operatorDescription
    every { mockToolSpec.getInputKeys() } returns setOf("input1", "input2")
    every { mockToolSpec.getOutputKeys() } returns setOf("output1")
  }

  @Test
  fun testGetOperatorByName() {
    // Given
    (registry.getOperatorsByName() as MutableMap)[operatorName] = mockOperator

    // When
    val result = registry.get(operatorName)

    // Then
    assertThat(result).isNotNull
    assertThat(result).isEqualTo(mockOperator)
  }

  @Test
  fun testGetOperatorByNameNotFound() {
    // When
    val result = registry.get("non-existent-operator")

    // Then
    assertThat(result).isNull()
  }

  @Test
  fun testSearchByTaskDescription() {
    // Given
    val taskDescription = "process some text"
    val operatorEmbedding = OperatorEmbedding(
      operatorName = operatorName,
      embedding = testEmbedding,
      hash = "test-hash",
    )

    (registry.getOperatorsByName() as MutableMap)[operatorName] = mockOperator

    every {
      embeddingRepository.findTopByEmbeddingSimilarity(
        any<FloatArray>(),
        any<Pageable>()
      )
    } returns listOf(operatorEmbedding)

    // When
    val results = registry.searchByTask(taskDescription)

    // Then
    assertThat(results).hasSize(1)
    assertThat(results.first()).isEqualTo(mockToolSpec)
    verify { embeddingService.generateEmbedding(taskDescription) }
    verify { embeddingRepository.findTopByEmbeddingSimilarity(testEmbedding, PageRequest.of(0, 3)) }
  }

  @Test
  fun testSearchByTaskDescriptionWithMaxResults() {
    // Given
    val taskDescription = "process some text"
    val operatorEmbedding = OperatorEmbedding(
      operatorName = operatorName,
      embedding = testEmbedding,
      hash = "test-hash",
    )

    (registry.getOperatorsByName() as MutableMap)[operatorName] = mockOperator

    every {
      embeddingRepository.findTopByEmbeddingSimilarity(
        any<FloatArray>(),
        any<Pageable>()
      )
    } returns listOf(operatorEmbedding)

    // When
    val results = registry.searchByTask(taskDescription)

    // Then
    assertThat(results).hasSize(1)
    assertThat(results.first()).isEqualTo(mockToolSpec)
    verify { embeddingRepository.findTopByEmbeddingSimilarity(testEmbedding, PageRequest.of(0, 3)) }
  }

  @Test
  fun testSearchByTaskHandlesException() {
    // Given
    val taskDescription = "process some text"
    every { embeddingService.generateEmbedding(taskDescription) } throws RuntimeException("Embedding service unavailable")

    // When
    val results = registry.searchByTask(taskDescription)

    // Then
    assertThat(results).isEmpty()
  }

  @Test
  fun testCreateEmbeddingText() {
    // Given - using package-private method for testing
    val embeddingText = registry.createEmbeddingTextForTesting(mockToolSpec)

    // Then - check individual components since Set order is not guaranteed
    assertThat(embeddingText)
      .contains("Tool: $operatorName")
      .contains(operatorDescription)
      .contains("Inputs:")
      .contains("input1")
      .contains("input2")
      .contains("Outputs: output1")
  }

  @Test
  fun testInitializeWithExistingEmbedding() {
    // Given
    every { embeddingRepository.findByOperatorName(operatorName) } returns Optional.of(
      OperatorEmbedding(
        operatorName = operatorName,
        embedding = testEmbedding,
        hash = "test-hash",
      )
    )
  }
}